docs/content/guides/dagster/building-and-memoizing-versioned-assets.mdx

```diff
@@ -0,0 +1,191 @@
+---
+title: Building and memoizing versioned assets | Dagster Docs
+---
+
+# Building and memoizing versioned assets
+
+This guide demonstrates how to build memoizable graphs of assets. Memoizable assets help avoid unnecessary recomputation, speed up the developer workflow, and save computational resources.
+
+---
+
+# Context
+
+Dagster embraces functional programming principles. The core abstraction is a software-defined asset which, in its idealized form, is a side-effect-free ("pure") function that produces an output asset from input assets.
+
+Pure functions can be made more efficient with a technique called "memoization". A memoized function has an outer layer in which previously computed outputs are cached and indexed by their corresponding inputs. When a memoized function is called twice with identical inputs, the cached result can be returned on the second call.
+
+In computationally expensive data pipelining, this approach can yield tremendous benefits. Dagster offers support for memoized assets via a versioning system. Because both code and input values can affect the output of a function, Dagster has two kinds of versions:
+
+1. Code version. A string that represents the version of the code that computes an asset. This is the `code_version` argument of <PyObject object="asset" decorator="true" />.
```

Comment:
per @sryza's feedback let's take a shot at this with more approachable language. maybe "cached"? or some variant

Reply:
My suggestion would be to deemphasize the language of memoization, caching, and functional programming and instead talk about what's happening in concrete terms.

It's worth throwing in a sentence to help people pattern-match that it's related to these ideas, but IMO we shouldn't open with them.

Here's a possible start:
---

There's no reason to spend time materializing an asset if the result is going to be the same as the result of its last materialization.

Dagster's versioning system helps you determine ahead of time whether materializing an asset will produce a different result. It's based on the idea that the result of an asset materialization shouldn't change as long as:
* The code used is the same code as the last time the asset was materialized.
* The input data is the same input data as the last time the asset was materialized.